### PR TITLE
feat: retry count & wait time objc

### PR DESCRIPTION
**Pull request by jayahariv**
* add property for retry count and retry wait time
* set the lite core replicator option

### CODE REVIEW

**Comment by jayahariv** [comment]
Looks like the failure of the test is not depended on this PR. Will check the failure, as a separate task. 

```

[2021-02-14T05:27:00.861Z] Failing tests:
[2021-02-14T05:27:00.861Z] 	CBL_EE_ObjC_Tests:
[2021-02-14T05:27:00.861Z] 		-[URLEndpointListenerTest testStopListener]
[2021-02-14T05:27:00.861Z] 
[2021-02-14T05:27:00.861Z] ** TEST FAILED **
```

File: Objective-C/CBLReplicatorConfiguration.m [hunk 1]
```diff
@@ -146,6 +152,30 @@ - (void) setHeartbeat: (NSTimeInterval)heartbeat {
     _heartbeat = heartbeat;
 }
 
+- (void) setMaxRetries: (NSInteger)maxRetries {
+    [self checkReadonly];
+    
+    if (maxRetries <= 0)
```
Comment (jayahariv) [inline Objective-C/CBLReplicatorConfiguration.m:158]:
```suggestion
    if (maxRetries < 0)
```

File: Objective-C/CBLReplicatorConfiguration.m [hunk 2]
```diff
@@ -146,6 +152,30 @@ - (void) setHeartbeat: (NSTimeInterval)heartbeat {
     _heartbeat = heartbeat;
 }
 
+- (void) setMaxRetries: (NSInteger)maxRetries {
+    [self checkReadonly];
+    
+    if (maxRetries < 0)
+        [NSException raise: NSInvalidArgumentException
+                    format: @"Attempt to store zero or negative value in maxRetries"];
```
Comment (pasin) [inline Objective-C/CBLReplicatorConfiguration.m:160]: "Attempt to store a negative value in maxRetries"

File: Objective-C/CBLReplicatorConfiguration.m [hunk 3]
```diff
@@ -215,6 +247,12 @@ - (NSDictionary*) effectiveOptions {
     if (_heartbeat > 0)
         options[@kC4ReplicatorHeartbeatInterval] = @(_heartbeat);
     
+    if (_maxRetryWaitTime > 0)
+        options[@kC4ReplicatorOptionMaxRetryInterval] = @(_maxRetryWaitTime);
+    
+    if (_maxRetries >= 0)
+        options[@kC4ReplicatorOptionMaxRetries] = @(_maxRetries);
```
Comment (pasin) [inline Objective-C/CBLReplicatorConfiguration.m:254]: _maxRetries is -1 by default so it needs to set to an appropriate value for continuous or single shot here.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLReplicatorConfiguration.m [hunk 4]
```diff
@@ -215,6 +247,12 @@ - (NSDictionary*) effectiveOptions {
     if (_heartbeat > 0)
         options[@kC4ReplicatorHeartbeatInterval] = @(_heartbeat);
     
+    if (_maxRetryWaitTime > 0)
+        options[@kC4ReplicatorOptionMaxRetryInterval] = @(_maxRetryWaitTime);
+    
+    if (_maxRetries >= 0)
+        options[@kC4ReplicatorOptionMaxRetries] = @([self maxRetries]);
```
Comment (pasin) [inline Objective-C/CBLReplicatorConfiguration.m:254]: `if (_maxRetries >= 0)` should be removed?

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLReplicatorConfiguration.m (hunk 4 above)
Comment (jayahariv) [reply Objective-C/CBLReplicatorConfiguration.m:254]: :(

**Review by pasin — APPROVED** [review]

**Pull request merged**